Modules/posixmodule.c

```diff
@@ -7252,13 +7225,40 @@ os_symlink_impl(PyObject *module, path_t *src, path_t *dst,
 
 #ifdef MS_WINDOWS
 
+    if (windows_has_symlink_unprivileged_flag) {
+        /* Allow non-admin symlinks if system allows it. */
+        flags |= SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE;
+    }
+
     Py_BEGIN_ALLOW_THREADS
-    /* if src is a directory, ensure target_is_directory==1 */
-    target_is_directory |= _check_dirW(src->wide, dst->wide);
-    result = Py_CreateSymbolicLinkW(dst->wide, src->wide,
-                                    target_is_directory);
+    /* if src is a directory, ensure flags==1 (target_is_directory bit) */
+    flags |= target_is_directory | _check_dirW(src->wide, dst->wide);
+
+    result = CreateSymbolicLinkW(dst->wide, src->wide, flags);
     Py_END_ALLOW_THREADS
 
+    if (windows_has_symlink_unprivileged_flag && !result) {
+        Py_BEGIN_ALLOW_THREADS
+        if (ERROR_INVALID_PARAMETER == GetLastError()) {
+            /* This error might be caused by
+            SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE not being supported.
+            Try again, and update windows_has_symlink_unprivileged_flag if we
+            are successful this time.
+
+            NOTE: There is a risk of a race condition here if there are other
+            conditions than the flag causing ERROR_INVALID_PARAMETER, and
+            another process (or thread) changes that condition in between our
+            calls to CreateSymbolicLink.
+            */
+            flags &= ~(SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE);
+            result = CreateSymbolicLinkW(dst->wide, src->wide, flags);
+            if (result || ERROR_INVALID_PARAMETER != GetLastError()) {
+                windows_has_symlink_unprivileged_flag = FALSE;
```

Comment:
Now `windows_has_symlink_unprivileged_flag` is modified with threads enabled again. Note that it is not necessary to call `GetLastError` within the region, so you may surround just the `CreateSymbolicLinkW` call with the macros. `Py_END_ALLOW_THREADS` is supposed to preserve the last error.

Reply:
Thanks for bearing with me in this. For future reference, are the nuances of `Py_BEGIN_ALLOW_THREADS` / `Py_END_ALLOW_THREADS` documented anywhere?

Reply:
The high-level documentation is in [Python C API docs](https://docs.python.org/3/c-api/init.html#thread-state-and-the-global-interpreter-lock). Regarding preservation of the last error, I've learned it from the source code, in particular, `PyThreadState_Swap` in Python/pystate.c and `PyThread_tss_get` in Python/thread_nt.h